server/webhook.go

```
@@ -246,6 +250,18 @@ func validateGitHubWebhookSignature256(signature, key string, body []byte) (bool
 	return subtle.ConstantTimeCompare([]byte(signature), []byte(got)) == 1, nil
 }
 
+// parseBranchNameFromRefs is for GitHub.
+// GitLab can configure branches that trigger push events, but GitHub cannot do it. So we need parse the branch name from refs field in request.
+// https://docs.github.com/en/rest/git/refs
+func parseBranchNameFromRefs(ref string) string {
+	expectedPrefix := "refs/heads/"
+	if !strings.HasPrefix(ref, expectedPrefix) {
```

If it's an exception path, we should return an error here.